#**BUSCANDO NUESTRO PÚBLICO OBJETIVO**
####Proyecto final de la cursada de Introducción a Análisis de Datos dictado por Alkemy.

###**INSTALACIÓN E IMPORTACIÓN DE LIBRERÍAS**
####Como primer paso debemos instalar e importar las librerías que vamos a utlizar:

In [1]:
!pip install pandas==2.2.2 #Instalamos la última versión de Pandas

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.0/13.0 MB 30.6 MB/s eta 0:00:00
  Attempting uninstall: pandas
    Found existing installation: pandas 2.0.3
    Uninstalling pandas-2.0.3:
      Successfully uninstalled pandas-2.0.3
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
cudf-cu12 24.4.1 requires pandas<2.2.2dev0,>=2.0, but you have pandas 2.2.2 which is incompatible.
google-colab 1.0.0 requires pandas==2.0.3, but you have pandas 2.2.2 which is incompatible.


In [2]:
import pandas as pd
import os

###**CARGA DE ARCHIVOS CSV ORIGINALES**
####Como segundo paso, vamos a realizar la carga de los archivos.csv que vamos a analizar en este proyecto. En este caso, lo vamos a hacer desde internet:

###*Customers*

In [3]:
#Leemos y creamos nuestro dataframe Customers a partir de un archivo .csv
df_Customers = pd.read_csv("https://raw.githubusercontent.com/DayQuiroga/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_customers_dataset.csv")

df_Customers.head() #Con el método .head visualizamos las primeras columnas de nuestro dataframe

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

###*Order items*

In [4]:
df_Order_items = pd.read_csv("https://raw.githubusercontent.com/DayQuiroga/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_order_items_dataset.csv")

df_Order_items.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

###*Payments*

In [5]:
df_Payments = pd.read_csv("https://raw.githubusercontent.com/DayQuiroga/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_order_payments_dataset.csv")

df_Payments.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

###*Orders*

In [7]:
df_Orders = pd.read_csv("https://raw.githubusercontent.com/DayQuiroga/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_orders_dataset.csv")

df_Orders.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

###*Products*

In [9]:
df_Products = pd.read_csv("https://raw.githubusercontent.com/DayQuiroga/recursos-analisis-datos/main/14_Datasets_Originales/ecommerce_products_dataset.csv")

df_Products.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

###**EXPLORACIÓN DE LOS DATAFRAMES**
####Como tercer paso realizaremos un análisis de la información contenida en nuestros dataframes, para poder comprender así su estructura, el tipo de datos que posee, si hay valores nulos o duplicados, etc.
####Para ello crearemos una función (exploracion_dataframe) que nos permita analizar cada uno de ellos:

In [16]:
def exploracion_dataframe(df):
    print("----------") #Los guiones están con el único objetivo de que al imprimirse la función, los resultados se muestren separados por los mismos
    print("Tamaño del DataFrame")
    print(df.shape) # Con .shape obtengo la cantidad de filas y columnas que posee mi dataframe
    print("----------")
    print("Tipo de datos")
    print(df.info()) # Con .info obtengo los nombres de las columnas y el tipo de datos que poseen, la cantidad de valores NO nulos que hay en cada una, etc.
    print("----------")
    print("Recuento de valores nulos")
    print(df.isnull().sum()) # Obtengo la cantidad de valores nulos que poseo en cada columna
    print("----------")
    print("Recuento de duplicados")
    print(df.duplicated().sum()) # Este método nos proporciona la cantidad de filas duplicadas
    print("----------")


####Una vez creada nuestra función, la aplicamos a cada uno de nuestros dataframes:

In [17]:
exploracion_dataframe(df_Customers)

----------
Tamaño del DataFrame
(99441, 5)
----------
Tipo de datos
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB
None
----------
Recuento de valores nulos
customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64
----------
Recuento de duplicados
0
----------


In [18]:
exploracion_dataframe(df_Order_items)

----------
Tamaño del DataFrame
(112650, 7)
----------
Tipo de datos
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB
None
----------
Recuento de valores nulos
order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64
----------
Recuento de duplicados
0
----------


In [19]:
exploracion_dataframe(df_Payments)

----------
Tamaño del DataFrame
(103886, 5)
----------
Tipo de datos
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB
None
----------
Recuento de valores nulos
order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64
----------
Recuento de duplicados
0
----------


In [20]:
exploracion_dataframe(df_Orders)

----------
Tamaño del DataFrame
(99441, 8)
----------
Tipo de datos
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB
None
----------
Recuento de valores nulos
order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0


In [21]:
exploracion_dataframe(df_Products)

----------
Tamaño del DataFrame
(32951, 9)
----------
Tipo de datos
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB
None
----------
Recuento de valores nulos
product_id                      0
product_category_name         610
product_name_lenght           610
p

####Como podemos ver gracias a esta exploración, los dataframes que poseen valores nulos son el de Orders y el de Products. Por ende, vamos a averiguar el porcentaje de nulos en cada uno de ellos:

####*Función porcentaje de nulos:*

In [38]:
def porcentaje_nulos (df):
   porcentaje_nulos = (df.isnull().sum() / len(df)) * 100 #Porcentaje de nulos en cada columna
   porcentaje_nulos = porcentaje_nulos.round(2) #Con .round redondeo decimales, en este caso a 2

   print(porcentaje_nulos)

In [39]:
porcentaje_nulos(df_Orders) #Aplicamos nuestra función porcentaje_nulos en el dataframe Orders

order_id                         0.00
customer_id                      0.00
order_status                     0.00
order_purchase_timestamp         0.00
order_approved_at                0.16
order_delivered_carrier_date     1.79
order_delivered_customer_date    2.98
order_estimated_delivery_date    0.00
dtype: float64


In [40]:
porcentaje_nulos(df_Products) #Aplicamos nuestra función porcentaje_nulos en el dataframe Products

product_id                    0.00
product_category_name         1.85
product_name_lenght           1.85
product_description_lenght    1.85
product_photos_qty            1.85
product_weight_g              0.01
product_length_cm             0.01
product_height_cm             0.01
product_width_cm              0.01
dtype: float64


###**TRANSFORMACIÓN DE LOS DATAFRAMES**
####Como pudimos ver en nuestra exploración de los dataframes, hay algunos de ellos que poseen valores nulos (los cuales se podrían eliminar en el caso de que su porcentaje no sea muy alto). También hay varias columnas que poseen formato object (por ejemplo, columnas que guardan fechas).
####Para poder realizar una mejor manipulación de nuestra información y evitar problemas a futuro, vamos a pasar las columnas de tipo object a *string* o a formato *fechas* (datetime64[ns]), eliminaremos las filas con valores nulos, y nos aseguraremos de que ciertas columnas estén completamente en *mayúscula* o en *minúscula* (para evitar que una palabra esté escrita con ambos tipos de letra, como por ejemplo Delivered en vez de delivered).

####*Función para transformar datos tipo object a string*:

In [22]:
def object_to_string(df, columnas_string): #Los parametros son el dataframe y la lista con las columnas a transformar
    for columna in columnas_string:
        df[columna] = df[columna].astype("string")
    return df

####*Función para transformar datos tipo object a fecha:*

In [66]:
def object_to_date(df, columnas_fechas): #Los parametros son el dataframe y la lista con las columnas a transformar
    df[columnas_fechas] = df[columnas_fechas].apply(pd.to_datetime)
    return df

####*Función para convertir minúsculas a mayúsculas:*

In [25]:
def minusc_to_mayusc(df, columna): #Los parametros son el dataframe y la lista con las columna a transformar
    df[columna] = df[columna].str.upper()
    return df


####*Función para corroborar si una columna se encuentra escrita en mayúscula  en su totalidad:*

In [58]:
def verificar_mayusculas(df, columna): #Los parámetros son el dataframe y la columna o lista de columnas del mismo que se quiere verificar
  return df[columna].str.isupper().all()

####**Transformamos nuestro dataframe Customers:**
####En este dataframe vamos a pasar todas las columnas a tipo string, y vamos a transformar las columnas customer_city y customer_state a mayúsculas.

In [42]:
columnas_customers_string = ["customer_id", "customer_unique_id", "customer_city", "customer_zip_code_prefix", "customer_state"] #Lista de columnas para transformar a strings

In [44]:
df_Customers = object_to_string(df_Customers, columnas_customers_string) #Aplicamos nuestra función object_to_string a la lista de columnas del df Customers a transformar

In [46]:
df_Customers.info() #Con .info corroboramos que ahora todas las columnas sean de tipo string

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  string
 1   customer_unique_id        99441 non-null  string
 2   customer_zip_code_prefix  99441 non-null  string
 3   customer_city             99441 non-null  string
 4   customer_state            99441 non-null  string
dtypes: string(5)
memory usage: 3.8 MB


In [59]:
df_Customers = minusc_to_mayusc(df_Customers, "customer_city") #Transformamos de minúscula a mayúscula con la función minusc_to_mayusc a la columna customer_city

####Ahora vamos a aplicar nuestra función verificar_mayusculas. Si el resultado de la misma es **True**, esto significa que todas las filas de la columna analizada están en mayúsculas. En el caso de ser **False**, entonces habría algunas columnas que no están en mayúsculas, por ende, habría que convertir la misma:

In [62]:
todo_mayuscula = verificar_mayusculas(df_Customers, "customer_state") #Verificamos si todas la columna está escrita en mayúscula
print(todo_mayuscula)

True


In [60]:
df_Customers.head() #Verificamos que la transformación se realizó correctamente

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

  customer_zip_code_prefix          customer_city customer_state  
0                    14409                 FRANCA             SP  
1                     9790  SAO BERNARDO DO CAMPO             SP  
2                     1151              SAO PAULO             SP  
3                     8775        MOGI DAS CRUZES             SP  
4                    13056               CAMPINAS             SP

####**Transformamos nuestro dataframe Order_items:**

In [63]:
columnas_Orderitems_string = ["order_id", "product_id", "seller_id"] #Lista de columnas para transformar de tipo object a string

In [64]:
df_Order_items = object_to_string(df_Order_items, columnas_Orderitems_string) #Aplicamos nuestra función object_to_string a la lista de columnas del df Order_items a transformar

In [68]:
df_Order_items = object_to_date(df_Order_items, "shipping_limit_date") #Aplicamos nuestra función object_to_date a la columna shipping_limit_date para que pase a tipo fecha

In [69]:
df_Order_items.info() #Con .info corroboramos que ahora todas las columnas sean del tipo de dato deseado

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   order_id             112650 non-null  string        
 1   order_item_id        112650 non-null  int64         
 2   product_id           112650 non-null  string        
 3   seller_id            112650 non-null  string        
 4   shipping_limit_date  112650 non-null  datetime64[ns]
 5   price                112650 non-null  float64       
 6   freight_value        112650 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(1), string(3)
memory usage: 6.0 MB


####**Transformamos nuestro dataframe Payments:**

In [71]:
columnas_payments_string = ["order_id", "payment_type"] #Lista de columnas para transformar de tipo object a string

In [73]:
df_Payments = object_to_string(df_Payments, columnas_payments_string) #Aplicamos nuestra función object_to_string a la lista de columnas del df Payments a transformar

In [74]:
df_Payments = minusc_to_mayusc(df_Payments, "payment_type") #Transformamos de minúscula a mayúscula con la función minusc_to_mayusc a la columna payment_type

In [76]:
df_Payments.info() #Corroboramos que el tipo de dato en cada columna sea el deseado

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  string 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  string 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), string(2)
memory usage: 4.0 MB


In [78]:
df_Payments.head() #Verificamos que la columna payment_type se encuentre en mayúsculas

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  CREDIT_CARD   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  CREDIT_CARD   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  CREDIT_CARD   
3  ba78997921bbcdc1373bb41e913ab953                   1  CREDIT_CARD   
4  42fdf880ba16b47b59251dd489d4441a                   1  CREDIT_CARD   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

####**Transformamos nuestro dataframe Orders:**

In [79]:
columnas_orders_string = ["order_id",	"customer_id", "order_status"] #Lista de columnas que pasarán de tipo objeto a string

In [80]:
df_Orders = object_to_string(df_Orders, columnas_orders_string) #Aplicamos nuestra función object_to_string a la lista de columnas del df Orders a transformar


In [81]:
columnas_orders_date = ["order_purchase_timestamp", "order_approved_at", "order_delivered_carrier_date", "order_delivered_customer_date", "order_estimated_delivery_date"] #Lista de columnas que pasarán de tipo object a fecha

In [82]:
df_Orders = object_to_date(df_Orders, columnas_orders_date) #Aplicamos nuestra función object_to_date a la lista de columnas del df Orders a transformar

In [83]:
df_Orders = minusc_to_mayusc(df_Orders, "order_status") #Transformamos de minúscula a mayúscula la columna order_status

####Como vimos durante la exploración de los dataframes, este posee valores nulos en algunas columnas. Como el porcentaje de los mismos es muy bajo, decidimos eliminar las filas que los contienen:

In [86]:
df_Orders.dropna(inplace=True) #Eliminamos las filas con valores nulos

In [88]:
df_Orders.shape #Verificamos que se eliminaron las filas con valores nulos

(96461, 8)

####Como podemos ver, las filas conteniendo nulos se han eliminado, ya que al utilizar el método .shape en mi dataframe Orders original, el resultado de el mismo era (99441, 8), o sea, el dataframe original poseía mayor cantidad de filas.

In [89]:
df_Orders.info() #Con .info corroboramos que ahora todas las columnas sean del tipo de dato deseado

<class 'pandas.core.frame.DataFrame'>
Index: 96461 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       96461 non-null  string        
 1   customer_id                    96461 non-null  string        
 2   order_status                   96461 non-null  string        
 3   order_purchase_timestamp       96461 non-null  datetime64[ns]
 4   order_approved_at              96461 non-null  datetime64[ns]
 5   order_delivered_carrier_date   96461 non-null  datetime64[ns]
 6   order_delivered_customer_date  96461 non-null  datetime64[ns]
 7   order_estimated_delivery_date  96461 non-null  datetime64[ns]
dtypes: datetime64[ns](5), string(3)
memory usage: 6.6 MB


In [90]:
df_Orders.tail() #Verificamos que el pasaje de minúscula a mayúscula de la columna order_status se haya realizado correctamente

order_id                       customer_id  \
99436  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99437  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99438  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99439  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99440  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp   order_approved_at  \
99436    DELIVERED      2017-03-09 09:54:05 2017-03-09 09:54:05   
99437    DELIVERED      2018-02-06 12:58:58 2018-02-06 13:10:37   
99438    DELIVERED      2017-08-27 14:46:43 2017-08-27 15:04:16   
99439    DELIVERED      2018-01-08 21:28:27 2018-01-08 21:36:21   
99440    DELIVERED      2018-03-08 20:57:30 2018-03-09 11:20:28   

      order_delivered_carrier_date order_delivered_customer_date  \
99436          2017-03-10 11:18:03           2017-03-17 15:08:01   
99437          2018-02-07 23:22:42           2018-02-28 17:37:56   
99438          2017-08-28 20:52:26           2017-09-21 11:24:17   
99439          2018-01-12 15:35:03           2018-01-25 23:32:54   
99440          2018-03-09 22:11:59           2018-03-16 13:08:30   

      order_estimated_delivery_date  
99436                    2017-03-28  
99437                    2018-03-02  
99438                    2017-09-27  
99439                    2018-02-15  
99440                    2018-04-03

In [92]:
df_Orders.isnull().sum() #Corroboramos si hay alguna columna con valores nulos

order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     0
order_delivered_customer_date    0
order_estimated_delivery_date    0
dtype: int64

####**Transformamos nuestro dataframe Products:**
####Como en este dataframe existen filas con valores nulos y el porcentaje de los mismos es bajo (como concluimos en nuestra sección de exploración de los dataframes), lo primero que realizaremos es la eliminación de las filas que los contienen:

In [93]:
df_Products.dropna(inplace=True) #Eliminamos las filas con valores nulos

In [94]:
df_Products.shape #Verificamos que se eliminaron las filas con valores nulos

(32340, 9)

####Como podemos ver, las filas conteniendo nulos se han eliminado, ya que al utilizar el método .shape en mi dataframe Products original, el resultado de el mismo era (32951, 9), o sea, el dataframe original poseía mayor cantidad de filas.

In [95]:
df_Products.isnull().sum() #Corroboramos si hay alguna columna con valores nulos

product_id                    0
product_category_name         0
product_name_lenght           0
product_description_lenght    0
product_photos_qty            0
product_weight_g              0
product_length_cm             0
product_height_cm             0
product_width_cm              0
dtype: int64

In [96]:
columnas_products_string = ["product_id", "product_category_name"] #Lista de columnas a transformar de tipo object a string

In [97]:
df_Products = object_to_string(df_Products, columnas_products_string) #Aplicamos nuestra función object_to_string a la lista de columnas del df Products a transformar

In [98]:
df_Products = minusc_to_mayusc(df_Products, "product_category_name") #Transformamos de minúscula a mayúscula la columna product_category_name

In [100]:
df_Products.info() #Con .info corroboramos que ahora todas las columnas sean del tipo de dato deseado

<class 'pandas.core.frame.DataFrame'>
Index: 32340 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32340 non-null  string 
 1   product_category_name       32340 non-null  string 
 2   product_name_lenght         32340 non-null  float64
 3   product_description_lenght  32340 non-null  float64
 4   product_photos_qty          32340 non-null  float64
 5   product_weight_g            32340 non-null  float64
 6   product_length_cm           32340 non-null  float64
 7   product_height_cm           32340 non-null  float64
 8   product_width_cm            32340 non-null  float64
dtypes: float64(7), string(2)
memory usage: 2.5 MB


In [101]:
df_Products.head() #Verificamos que la columna product_category_name se encuentre en mayúsculas

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             PERFUMARIA   
1  3aa071139cb16b67ca9e5dea641aaa2f                  ARTES   
2  96bd76ec8810374ed1b65e291975717f          ESPORTE_LAZER   
3  cef67bcfe19066a932b7673e239eb23d                  BEBES   
4  9dc1a7de274444849c219cff195d0b71  UTILIDADES_DOMESTICAS   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

###**ANÁLISIS DE LOS DATAFRAMES**
####Ahora realizaremos un análisis de nuestros dataframes ya transformados, para poder sacar la mayor cantidad de datos relevantes de los mismos (*insights*).
####Para llevar esto a cabo, utilizaremos la función *analisisDF*:

In [103]:
def analisisDF(df): #El parámetro de esta función son los dataframes a analizar
    print("--------") #Los guiones están con el único objetivo de que al imprimirse la función, los resultados se muestren separados por los mismos
    print("Analisis de variables cuantitativas")
    columnas_numericas = df.select_dtypes(include=['int', 'float']).columns #Selecciono todas las columnas del dataframe que contienen datos numéricos y obtengo el nombre de estas columnas
    if not columnas_numericas.empty: #Verifico si el df tiene columnas de tipo numéricas y si las hay, se imprime un resumen descriptivo de las mismas con el método .describe()
        print(df[columnas_numericas].describe())
    print("--------")
    print("Analisis de variables cualitativas")
    columnas_string = df.select_dtypes(include=['string']).columns # #Selecciono todas las columnas del df que contienen datos tipo string y obtengo el nombre de estas columnas
    if not columnas_string.empty: #Verifico si el df tiene columnas de tipo string y si las hay, se imprime un resumen descriptivo de las mismas con el método .describe()
        print(df[columnas_string].describe())

####**Analizamos nuestro dataframe Customers:**

In [104]:
analisisDF(df_Customers) #Aplicamos la función analisisDF

--------
Analisis de variables cuantitativas
--------
Analisis de variables cualitativas
                             customer_id                customer_unique_id  \
count                              99441                             99441   
unique                             99441                             96096   
top     06b8999e2fba1a1fbc88172c00ba8bc7  8d50f5eadf50201ccdcedfb9e2ac8455   
freq                                   1                                17   

       customer_zip_code_prefix customer_city customer_state  
count                     99441         99441          99441  
unique                    14994          4119             27  
top                       22790     SAO PAULO             SP  
freq                        142         15540          41746  


In [109]:
#Realizamos un analisis por estado de los compradores
df_Customers["customer_state"].value_counts()

customer_state
SP    41746
RJ    12852
MG    11635
RS     5466
PR     5045
SC     3637
BA     3380
DF     2140
ES     2033
GO     2020
PE     1652
CE     1336
PA      975
MT      907
MA      747
MS      715
PB      536
PI      495
RN      485
AL      413
SE      350
TO      280
RO      253
AM      148
AC       81
AP       68
RR       46
Name: count, dtype: Int64

#####**Podemos ver que los estados con mayor cantidad de clientes son:**
*   Sao Pablo
*   Rio de Janeiro
*   Minas Gerais
#####**Los estados con menor cantidad de clientes son:**
*   Acre
*   Amapá
*   Roraima









####**Analizamos nuestro dataframe Order_items:**

In [105]:
analisisDF(df_Order_items) #Aplicamos la función analisisDF

--------
Analisis de variables cuantitativas
       order_item_id          price  freight_value
count  112650.000000  112650.000000  112650.000000
mean        1.197834     120.653739      19.990320
std         0.705124     183.633928      15.806405
min         1.000000       0.850000       0.000000
25%         1.000000      39.900000      13.080000
50%         1.000000      74.990000      16.260000
75%         1.000000     134.900000      21.150000
max        21.000000    6735.000000     409.680000
--------
Analisis de variables cualitativas
                                order_id                        product_id  \
count                             112650                            112650   
unique                             98666                             32951   
top     8272b63d03f5f79c56e9e4120aec44ef  aca2eb7d00ea1a7b8ebd4e68314663af   
freq                                  21                               527   

                               seller_id  
count             

*   Podemos ver que el promedio de los precios de todos los productos es de 120 reales.

*   El promedio del costo de envío es de 20 reales aproximadamente.
*   El vendedor con ID = 6560211a19b47992c3666cc44a7e94c0 fue el que más ha vendido en la plataforma. Esto se podría usar para ofrecerle algún beneficio al mismo e incentivarlo a que siga ofreciendo sus productos en la plataforma; o para luego averiguar que productos ofrece, y así buscar otros vendedores que ofrezcan productos similares y ofrecerles que se unan a vender en el ecommerce, para así obtener más ventas y que los clientes encuentren mayor variedad de estos productos en la plataforma.





####**Analizamos nuestro dataframe Payments:**

In [106]:
analisisDF(df_Payments) #Aplicamos la función analisisDF

--------
Analisis de variables cuantitativas
       payment_sequential  payment_installments  payment_value
count       103886.000000         103886.000000  103886.000000
mean             1.092679              2.853349     154.100380
std              0.706584              2.687051     217.494064
min              1.000000              0.000000       0.000000
25%              1.000000              1.000000      56.790000
50%              1.000000              1.000000     100.000000
75%              1.000000              4.000000     171.837500
max             29.000000             24.000000   13664.080000
--------
Analisis de variables cualitativas
                                order_id payment_type
count                             103886       103886
unique                             99440            5
top     fa65dad1b0e818e3ccc5cb0e39231352  CREDIT_CARD
freq                                  29        76795


In [110]:
#Investigamos tipos de pago
df_Payments["payment_type"].value_counts()

payment_type
CREDIT_CARD    76795
BOLETO         19784
VOUCHER         5775
DEBIT_CARD      1529
NOT_DEFINED        3
Name: count, dtype: Int64

*   Podemos ver que el medio de pago mayormente utilizado es la tarjeta de crédito. Saber esto nos podría servir para ofrecer algún tipo de beneficio a nuestros clientes si utlizan este método de pago, y así atraer más ventas.




####**Analizamos nuestro dataframe Orders:**

In [107]:
analisisDF(df_Orders) #Aplicamos la función analisisDF

--------
Analisis de variables cuantitativas
--------
Analisis de variables cualitativas
                                order_id                       customer_id  \
count                              96461                             96461   
unique                             96461                             96461   
top     e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
freq                                   1                                 1   

       order_status  
count         96461  
unique            2  
top       DELIVERED  
freq          96455  


In [114]:
#Observo el estado de los envios
df_Orders["order_status"].value_counts()

order_status
DELIVERED    96455
CANCELED         6
Name: count, dtype: Int64

*   Podemos ver que casi todos los pedidos han sido entregados (delivered).




In [118]:
#Calculo el porcentaje de pedidos entregados
porcentaje_delivered = (df_Orders['order_status'] == 'DELIVERED').mean() * 100
porcentaje_delivered

99.9937798695846

*   Vemos que nuestro porcentaje de pedidos entregados es altísimo (99,9%), por lo tanto, concluimos que el proceso de entrega es óptimo.




####**Analizamos nuestro dataframe Products:**

In [108]:
analisisDF(df_Products) #Aplicamos la función analisisDF

--------
Analisis de variables cuantitativas
       product_name_lenght  product_description_lenght  product_photos_qty  \
count         32340.000000                32340.000000        32340.000000   
mean             48.476592                  771.492393            2.188961   
std              10.245699                  635.124831            1.736787   
min               5.000000                    4.000000            1.000000   
25%              42.000000                  339.000000            1.000000   
50%              51.000000                  595.000000            1.000000   
75%              57.000000                  972.000000            3.000000   
max              76.000000                 3992.000000           20.000000   

       product_weight_g  product_length_cm  product_height_cm  \
count      32340.000000       32340.000000       32340.000000   
mean        2276.956586          30.854545          16.958813   
std         4279.291845          16.955965          13.63

In [119]:
#Analizamos la columna category_name
df_Products["product_category_name"].value_counts()

product_category_name
CAMA_MESA_BANHO                  3029
ESPORTE_LAZER                    2867
MOVEIS_DECORACAO                 2657
BELEZA_SAUDE                     2444
UTILIDADES_DOMESTICAS            2335
                                 ... 
FASHION_ROUPA_INFANTO_JUVENIL       5
CASA_CONFORTO_2                     5
PC_GAMER                            3
SEGUROS_E_SERVICOS                  2
CDS_DVDS_MUSICAIS                   1
Name: count, Length: 73, dtype: Int64

Podemos ver que las 4 categorías de productos con mayor cantidad de ventas son, en orden decreciente:


*   CAMA_MESA_BANHO (cama_mesa_baño, con 3029 ventas)
*   ESPORTE_LAZER (deporte_ocio, con 2867 ventas)
*   MOVEIS_DECORACAO (muebles_decoración, con 2657 ventas)
*   BELEZA_SAUDE (belleza_salud, con 2444 ventas)

Saber esto podría servir para hacer más publicidad a estos tipos de productos, y así obtener más ventas.







###**NUEVOS DATAFRAMES**
####Luego de todo el trabajo realizado sobre nuestros dataframes originales, generamos dos nuevos dataframes (en formato.csv), los cuales luego podemos utilizar para generar otros reportes con, por ejemplo, distintas herramientas gráficas. Estos df se exportarán a la carpeta dataset_generados la cual fue creada manualmente.

####Realizamos un *merge* (combinación) entre los dataframes Orders y Payments:

In [120]:
df_orders_payments = pd.merge(df_Orders, df_Payments, how= "inner", on= "order_id")
df_orders_payments.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp   order_approved_at  \
0    DELIVERED      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    DELIVERED      2017-10-02 10:56:33 2017-10-02 11:07:15   
2    DELIVERED      2017-10-02 10:56:33 2017-10-02 11:07:15   
3    DELIVERED      2018-07-24 20:41:37 2018-07-26 03:24:27   
4    DELIVERED      2018-08-08 08:38:49 2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date  payment_sequential payment_type  \
0                    2017-10-18                   1  CREDIT_CARD   
1                    2017-10-18                   3      VOUCHER   
2                    2017-10-18                   2      VOUCHER   
3                    2018-08-13                   1       BOLETO   
4                    2018-09-04                   1  CREDIT_CARD   

   payment_installments  payment_value  
0                     1          18.12  
1                     1           2.00  
2                     1          18.59  
3                     1         141.46  
4                     3         179.12

In [121]:
#Elimino las columnas que no se utilizarán
df_orders_payments.drop(['customer_id','order_approved_at','order_delivered_carrier_date','order_delivered_customer_date','order_estimated_delivery_date'], axis=1, inplace=True) #Indico el nombre de las columnas que voy a mantener los cambios
df_orders_payments.head()

order_id order_status order_purchase_timestamp  \
0  e481f51cbdc54678b7cc49136f2d6af7    DELIVERED      2017-10-02 10:56:33   
1  e481f51cbdc54678b7cc49136f2d6af7    DELIVERED      2017-10-02 10:56:33   
2  e481f51cbdc54678b7cc49136f2d6af7    DELIVERED      2017-10-02 10:56:33   
3  53cdb2fc8bc7dce0b6741e2150273451    DELIVERED      2018-07-24 20:41:37   
4  47770eb9100c2d0c44946d9cf07ec65d    DELIVERED      2018-08-08 08:38:49   

   payment_sequential payment_type  payment_installments  payment_value  
0                   1  CREDIT_CARD                     1          18.12  
1                   3      VOUCHER                     1           2.00  
2                   2      VOUCHER                     1          18.59  
3                   1       BOLETO                     1         141.46  
4                   1  CREDIT_CARD                     3         179.12

In [122]:
#Exportamos como csv el dataframe
df_orders_payments.to_csv("dataset_generados/orders_payments.csv")

####Realizamos un *merge* entre los dataframes Order items y Products:

In [123]:
df_order_items_products = pd.merge(df_Order_items, df_Products, how = "inner", on= "product_id")
df_order_items_products.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

  shipping_limit_date   price  freight_value product_category_name  \
0 2017-09-19 09:45:35   58.90          13.29            COOL_STUFF   
1 2017-05-03 11:05:13  239.90          19.93              PET_SHOP   
2 2018-01-18 14:48:30  199.00          17.87      MOVEIS_DECORACAO   
3 2018-08-15 10:10:18   12.99          12.79            PERFUMARIA   
4 2017-02-13 13:57:51  199.90          18.14    FERRAMENTAS_JARDIM   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 58.0                       598.0                 4.0   
1                 56.0                       239.0                 2.0   
2                 59.0                       695.0                 2.0   
3                 42.0                       480.0                 1.0   
4                 59.0                       409.0                 1.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             650.0               28.0                9.0              14.0  
1           30000.0               50.0               30.0              40.0  
2            3050.0               33.0               13.0              33.0  
3             200.0               16.0               10.0              15.0  
4            3750.0               35.0               40.0              30.0

In [129]:
#Exportamos como csv nuestro nuevo dataframe
df_order_items_products.to_csv("dataset_generados/order_items_products.csv")